server/shared/src/main/scala/org/http4s/server/middleware/GZip.scala

```diff
@@ -89,4 +104,16 @@ object GZip extends GZipPlatform {
       .putHeaders(`Content-Encoding`(ContentCoding.gzip))
       .pipeBodyThrough(compressPipe)
   }
+
+  private def unzipRequest[F[_]: Compression](
+      bufferSize: Int,
+      req: Request[F],
+  ): Request[F] = {
+    val decompressPipe = Compression[F].gunzip(bufferSize = bufferSize).andThenF(_.content)
```

Comment:
I think the `.handleErrorWith` in https://github.com/http4s/http4s/pull/5546#issuecomment-980396914 is a good idea to throw `MalformedMessageBodyFailure` for streams that aren't valid gzip streams, but maybe it could be more specific (and I should have added a test for it).